File: jianzhioffer/firstround/jianzhi_32_2.py
import collections

# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        if not root:
            return []
        nodelist = []
        queue = collections.deque()
        queue.append(root)
        while queue:
            tmp = []
            for _ in range(len(queue)):
                node = queue.popleft()
                tmp.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            nodelist.append(tmp)
        return nodelist


if __name__ == '__main__':
    l1 = TreeNode(3)
    l2 = TreeNode(9)
    l3 = TreeNode(20)
    l4 = TreeNode(15)
    l5 = TreeNode(7)
    # l6 = TreeNode(4)
    # l7 = TreeNode(3)

    l1.left = l2
    l1.right = l3
    l3.left = l4
    l3.right = l5
    # l2.left = l4
    # l2.right = l5
    # l3.left = l6
    # l3.right = l7
    obj = Solution()
    result = obj.levelOrder(l1)
    print(result)

File: jianzhioffer/secondround/jianzhi_32_2.cpp
<<<<<<< HEAD
 #include <vector>
 #include <queue>
 using namespace std;

 //Definition for a binary tree node.
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

//BFS，时间复杂度O(N)，空间复杂度O(N)
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(root==NULL) return res;
        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty())
        {
            int current_size = que.size();
            vector<int> layer_nums;
            for (int i = 0; i < current_size; i++)
            {
                TreeNode* node = que.front();
                que.pop();
                layer_nums.push_back(node->val);
                if(node->left!=NULL) que.push(node->left);
                if(node->right!=NULL) que.push(node->right);
            }
            res.push_back(layer_nums);
        }
        return res;
    }
};

int main(int argc, char const *argv[])
{
    TreeNode t1 = TreeNode(1);
    TreeNode t2 = TreeNode(2);
    TreeNode t3 = TreeNode(3);
    // TreeNode t4 = TreeNode(4);
    // TreeNode t5 = TreeNode(5);
    TreeNode t6 = TreeNode(6);
    TreeNode t7 = TreeNode(7);

    t1.left = &t2;
    t1.right = &t3;
    // t2.left = &t4;
    // t2.right = &t5;
    t3.left = &t6;
    t3.right = &t7;

    Solution obj;
    vector<vector<int>> result = obj.levelOrder(&t1);
    return 0;
=======
 #include <vector>
 #include <queue>
 using namespace std;

 //Definition for a binary tree node.
 struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

//BFS，时间复杂度O(N)，空间复杂度O(N)
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> res;
        if(root==NULL) return res;
        queue<TreeNode*> que;
        que.push(root);
        while (!que.empty())
        {
            int current_size = que.size();
            vector<int> layer_nums;
            for (int i = 0; i < current_size; i++)
            {
                TreeNode* node = que.front();
                que.pop();
                layer_nums.push_back(node->val);
                if(node->left!=NULL) que.push(node->left);
                if(node->right!=NULL) que.push(node->right);
            }
            res.push_back(layer_nums);
        }
        return res;
    }
};

int main(int argc, char const *argv[])
{
    TreeNode t1 = TreeNode(1);
    TreeNode t2 = TreeNode(2);
    TreeNode t3 = TreeNode(3);
    // TreeNode t4 = TreeNode(4);
    // TreeNode t5 = TreeNode(5);
    TreeNode t6 = TreeNode(6);
    TreeNode t7 = TreeNode(7);

    t1.left = &t2;
    t1.right = &t3;
    // t2.left = &t4;
    // t2.right = &t5;
    t3.left = &t6;
    t3.right = &t7;

    Solution obj;
    vector<vector<int>> result = obj.levelOrder(&t1);
    return 0;
>>>>>>> c02bd06644e6a77d7bfda85f1876d137906a27bd
}